modules/core/include/opencv2/core/utility.hpp

```diff
@@ -513,30 +513,36 @@ class CV_EXPORTS AutoLock
     AutoLock& operator = (const AutoLock&);
 };
 
+// TLS interface
 class CV_EXPORTS TLSDataContainer
 {
-private:
-    int key_;
 protected:
-    TLSDataContainer();
-    virtual ~TLSDataContainer();
-public:
-    virtual void* createDataInstance() const = 0;
-    virtual void deleteDataInstance(void* data) const = 0;
+    TLSDataContainer() { key = (size_t)-1; };
+    virtual ~TLSDataContainer() {};
 
+    void  getKey();
     void* getData() const;
+    void  releaseKey();
```

Comment:
Please rename "releaseKey" into "release".
Also "release" is not needed in interface (for end-user), be we can't move implementation into destructor, because "release" calls virtual methods from derived class (which become unavailable from base class dtor).

TLSDataContainer dtor can't call "release", it should be done in derived classes:
```
TLSDataContainer::~TLSDataContainer()
{
    CV_Assert(key_ == -1 &&
"Please recompile application with updated OpenCV sources or check that 'release' method is called from destructor of derived class");
}
```
and below:
```
inline TLSData() { /* nothing */ }
inline ~TLSData() { release(); } // Release key and delete associated data
```